### PR TITLE
Remove span builder

### LINKED ISSUE #145: Define API to create span from the current Activity
In auto-collectors Activity is created by the library (HTTP client or web framework or anything else). We need to create Span that represents the same Activity - not create a child of it.

In #140  it looks like this

```csharp
ISpan span = this.Tracer.SpanBuilder(path)
  .SetSpanKind(SpanKind.Server)
  .SetCreateChild(false)
  .StartSpan();
```

`SetCreateChild(false)` does the trick, but does not seem to be good choice. We need to revisit this API and find better name/place We are going to work more on this integration and should have better ideas.

Any ideas are welcome

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #145 

SpanBuilder requires extra allocation for every Span created, it also makes span creation very verbose.

This PR suggest how to remove builder and create span directly with a lot of overloads on Tracer.

**Before**
```csharp
var span = tracer
    .SpanBuilder("outgoing http request")
    .SetSpanKind(SpanKind.Client)
    .StartSpan();
```

or 

```csharp
    var span = tracer
        .SpanBuilder("GET api/values") 
        .SetCreateChild(false) 
        .StartSpan();
```

**After**

```csharp
var span = tracer.StartSpan("outgoing http request", SpanKind.Client);
```

```csharp
var span = tracer.StartSpanFromActivity("GET api/values", Activity.Current)
```

## Benchmarks
- Noop allocations are gone, time to create noop span got twice as small
- for real span creation, there a slight decrease in allocations and time

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-4790 CPU 3.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview8-013656
  [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT

### Before
|                           Method |        Mean |      Error |     StdDev |  Ratio | RatioSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated |
|--------------------------------- |------------:|-----------:|-----------:|-------:|--------:|-------:|-------:|-------:|----------:|
|               CreateSpan_Sampled |   835.12 ns | 15.7607 ns | 13.1609 ns |  67.59 |    3.12 | 0.1984 |      - |      - |     840 B |
|    CreateSpan_Attributes_Sampled | 1,054.59 ns | 13.8295 ns | 12.9361 ns |  85.64 |    3.59 | 0.3624 |      - |      - |    1528 B |
|     CreateSpan_Propagate_Sampled | 1,893.55 ns | 31.8505 ns | 29.7930 ns | 153.75 |    6.36 | 0.2804 | 0.0038 | 0.0019 |    1207 B |
| CreateSpan_Attributes_NotSampled |   871.99 ns | 17.2950 ns | 33.7326 ns |  72.29 |    4.57 | 0.1993 |      - |      - |     840 B |
|                  CreateSpan_Noop |    12.16 ns |  0.2692 ns |  0.4269 ns |   1.00 |    0.00 | 0.0057 |      - |      - |      24 B |
|       CreateSpan_Attributes_Noop |    22.32 ns |  0.3673 ns |  0.3436 ns |   1.81 |    0.07 | 0.0057 |      - |      - |      24 B |
|        CreateSpan_Propagate_Noop |    14.79 ns |  0.2758 ns |  0.2580 ns |   1.20 |    0.05 | 0.0057 |      - |      - |      24 B |

### After
|                           Method |         Mean |      Error |     StdDev |  Ratio | RatioSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated |
|--------------------------------- |-------------:|-----------:|-----------:|-------:|--------:|-------:|-------:|-------:|----------:|
|               CreateSpan_Sampled |   803.906 ns | 15.9024 ns | 16.3306 ns | 131.65 |    3.78 | 0.1402 |      - |      - |     592 B |
|    CreateSpan_Attributes_Sampled | 1,032.477 ns | 25.2346 ns | 36.9885 ns | 171.31 |    6.92 | 0.3033 |      - |      - |    1280 B |
|     CreateSpan_Propagate_Sampled | 1,800.736 ns | 42.5193 ns | 53.7733 ns | 297.26 |   12.51 | 0.2232 | 0.0057 | 0.0038 |     972 B |
| CreateSpan_Attributes_NotSampled |   800.006 ns |  8.5102 ns |  7.9605 ns | 131.25 |    2.95 | 0.1402 |      - |      - |     592 B |
|                  CreateSpan_Noop |     6.097 ns |  0.1240 ns |  0.1160 ns |   1.00 |    0.00 |      - |      - |      - |         - |
|       CreateSpan_Attributes_Noop |    15.322 ns |  0.3289 ns |  0.5217 ns |   2.52 |    0.10 |      - |      - |      - |         - |
|        CreateSpan_Propagate_Noop |     7.787 ns |  0.0833 ns |  0.0779 ns |   1.28 |    0.03 |      - |      - |      - |         - |

### CODE REVIEW

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 1]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+
+        public ISpan StartSpan(string operationName)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName})");
+            return new LoggingSpan(operationName, SpanKind.Internal);
+        }
+
+        public ISpan StartSpan(string operationName, SpanKind kind)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName} {kind})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {kind})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 2]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+
+        public ISpan StartRootSpan(string operationName)
+        {
+            Logger.Log($"{prefix}.StartRootSpan({operationName})");
+            return new LoggingSpan(operationName, SpanKind.Internal);
+        }
+
+        public ISpan StartRootSpan(string operationName, SpanKind kind)
+        {
+            Logger.Log($"{prefix}.StartRootSpan({operationName} {kind})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartRootSpan({operationName}, {kind})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 3]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+
+        public ISpan StartSpan(string operationName, SpanKind kind, DateTimeOffset startTimestamp)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {kind}, {startTimestamp:o})");
+            return new LoggingSpan(operationName, kind);
+        }
+
+        public ISpan StartSpan(string operationName, SpanKind kind, DateTimeOffset startTimestamp, IEnumerable<Link> links)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {kind}, {startTimestamp:o} {links})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {kind}, {startTimestamp:o}, {links})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 4]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+
+        public ISpan StartSpan(string operationName, ISpan parent, SpanKind kind)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind})");
+            return new LoggingSpan(operationName, kind);
+        }
+
+        public ISpan StartSpan(string operationName, ISpan parent, SpanKind kind, DateTimeOffset startTimestamp)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind} {startTimestamp:o})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind}, {startTimestamp:o})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 5]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+            return new LoggingSpan(operationName, kind);
+        }
+
+        public ISpan StartSpan(string operationName, ISpan parent, SpanKind kind, DateTimeOffset startTimestamp, IEnumerable<Link> links)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind} {startTimestamp:o}, {links})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind}, {startTimestamp:o}, {links})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 6]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+
+        public ISpan StartSpan(string operationName, in SpanContext parent)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name})");
+            return new LoggingSpan(operationName, SpanKind.Internal);
+        }
+
+        public ISpan StartSpan(string operationName, in SpanContext parent, SpanKind kind)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name} {kind})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 7]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+            return new LoggingSpan(operationName, kind);
+        }
+
+        public ISpan StartSpan(string operationName, in SpanContext parent, SpanKind kind, DateTimeOffset startTimestamp)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name} {kind} {startTimestamp:o})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind}, {startTimestamp:o})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 8]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+            return new LoggingSpan(operationName, kind);
+        }
+
+        public ISpan StartSpan(string operationName, in SpanContext parent, SpanKind kind, DateTimeOffset startTimestamp,
+            IEnumerable<Link> links)
+        {
+            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name} {kind} {startTimestamp:o} {links})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpan({operationName}, {parent.GetType().Name}, {kind}, {startTimestamp:o}, {links})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 9]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+
+        public ISpan StartSpanFromActivity(string operationName, Activity activity)
+        {
+            Logger.Log($"{prefix}.StartSpanFromActivity({operationName}, {activity.OperationName})");
+            return new LoggingSpan(operationName, SpanKind.Internal);
+        }
+
+        public ISpan StartSpanFromActivity(string operationName, Activity activity, SpanKind kind)
+        {
+            Logger.Log($"{prefix}.StartSpanFromActivity({operationName}, {activity.OperationName} {kind})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpanFromActivity({operationName}, {activity.OperationName}, {kind})");
```

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 10]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
+            return new LoggingSpan(operationName, kind);
+        }
+
+        public ISpan StartSpanFromActivity(string operationName, Activity activity, SpanKind kind, IEnumerable<Link> links)
+        {
+            Logger.Log($"{prefix}.StartSpanFromActivity({operationName}, {activity.OperationName} {kind} {links})");
```
Comment (bruno-garcia) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]:
```suggestion
            Logger.Log($"{prefix}.StartSpanFromActivity({operationName}, {activity.OperationName}, {kind}, {links})");
```

File: src/OpenTelemetry/Trace/SpanBuilder.cs [hunk 11]
```diff
@@ -26,8 +26,7 @@ namespace OpenTelemetry.Trace
     using OpenTelemetry.Trace.Export;
     using OpenTelemetry.Trace.Internal;
 
-    /// <inheritdoc/>
-    public class SpanBuilder : ISpanBuilder
+    /*public class SpanBuilder
```
Comment (bruno-garcia) [inline src/OpenTelemetry/Trace/SpanBuilder.cs]: can we delete this file altogether?

File: src/OpenTelemetry.Collector.StackExchangeRedis/OpenTelemetry.Collector.StackExchangeRedis.csproj [hunk 12]
```diff
@@ -6,7 +6,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\OpenTelemetry.Abstractions\OpenTelemetry.Abstractions.csproj" />
+    <ProjectReference Include="..\OpenTelemetry\OpenTelemetry.csproj" />
```
Comment (pcwiese) [inline src/OpenTelemetry.Collector.StackExchangeRedis/OpenTelemetry.Collector.StackExchangeRedis.csproj]: I thought collectors should only references the API?

File: src/OpenTelemetry/Trace/Span.cs [hunk 13]
```diff
@@ -328,6 +332,12 @@ public void End()
 
         public void End(DateTimeOffset endTimestamp)
         {
+            if (this.hasEnded)
```
Comment (pcwiese) [inline src/OpenTelemetry/Trace/Span.cs]: Can you clarify the use of your lock? Seems like we should be protecting all public methods that can mutate the state of the Span. Two threads could end up at line 340 simultaneously...

Also, maybe consider changing the name of @lock to something that isn't an important keyword.

File: src/OpenTelemetry.Abstractions/Trace/ITracer.cs [hunk 14]
```diff
@@ -47,10 +49,184 @@ public interface ITracer
...
         /// <summary>
-        /// Gets the span builder for the span with the given name.
+        /// Creates root span.
         /// </summary>
-        /// <param name="spanName">Span name.</param>
-        /// <returns>Span builder for the span with the given name.</returns>
-        ISpanBuilder SpanBuilder(string spanName);
+        /// <param name="operationName">Span name.</param>
+        /// <returns>Span instance.</returns>
+        ISpan StartRootSpan(string operationName);
```
Comment (pcwiese) [inline src/OpenTelemetry.Abstractions/Trace/ITracer.cs]: Can many of these methods on ITracer be turned into extension methods instead? Most of the implementations in Tracer.cs are redirections to other ITracer methods.

**Comment by bogdandrutu** [comment]
Alternative option is to have an "StartOption" class for the optional fields. So you do have the "required" fields like name explicit and everything else in the Option. This will reduce the API surface and will achieve the performance benefit for the 90% of the cases.

File: src/OpenTelemetry.Abstractions/Trace/ITracer.cs (hunk 14 above)
Comment (lmolkova) [reply src/OpenTelemetry.Abstractions/Trace/ITracer.cs]: this is great idea, I think it should be possible to keep all extensions in abstractions without the need to override them in SDK. I'll try to explore this path

**Comment by lmolkova** [comment]
> Alternative option is to have an "StartOption" class for the optional fields. So you do have the "required" fields like name explicit and everything else in the Option. This will reduce the API surface and will achieve the performance benefit for the 90% of the cases.

The problem with small class is the same as builder - it's an allocation on heap. Which is exactly we what to avoid by removing builder.

The option is to have struct Options (stack allocation), but this struct will be too big and compicates usage. Personally I like how this API looks now from usage perspective (not internally though but this is lesser priority)

**Comment by lmolkova** [comment]
> Overall in favor, but we should be aware of [open-telemetry/opentelemetry-specification#238](https://github.com/open-telemetry/opentelemetry-specification/issues/238) and [open-telemetry/opentelemetry-specification#287](https://github.com/open-telemetry/opentelemetry-specification/pull/287); `StartSpan` shouldn't necessarily activate the span.

@austinlparker  Thanks for sharing! I've gone through the issues and there seems to be a consensus that there is start and activate (make current). It seems there are no plans to separate *create* and *start*. Did I miss anything? 

Just to clarify: we still have start and tracer.WithSpan separate. This PR merges create + start. And if someone wants delayed start, they could also delay create in the same way.

File: samples/LoggingTracer/LoggingTracer/LoggingTracer.cs [hunk 15]
```diff
@@ -23,16 +27,125 @@ internal LoggingTracer()
...
-            return new LoggingSpanBuilder(spanName, SpanKind.Internal);
-        }
-
         public IDisposable WithSpan(ISpan span)
         {
             Logger.Log($"{prefix}.WithSpan");
             return new CurrentSpanUtils.LoggingScope(span);
         }
+
+        public ISpan StartRootSpan(string operationName)
```
Comment (SergeyKanzhelev) [inline samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]: now all links needs to be set in "builder". So we will need more overloads

Comment (SergeyKanzhelev) [reply samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#add-links

Comment (lmolkova) [reply samples/LoggingTracer/LoggingTracer/LoggingTracer.cs]: you can use overloads that dop not have links - you don't have to set them if you don't have them.

we have all methods having overload with links. 

We'll also need to do lazy links - this is in TODO

File: src/OpenTelemetry/Trace/Span.cs (hunk 13 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Span.cs]: I agree the lock existence here is questionable and name could be improved. I'd prefer to address this separately from this PR - it's too big already  :) Create https://github.com/open-telemetry/opentelemetry-dotnet/issues/271 to follow up

**Pull request merged**